tests/ui/codegen/equal-pointers-unequal/as-cast/segfault.rs

```diff
@@ -0,0 +1,63 @@
+//@ known-bug: #107975
+//@ compile-flags: -Copt-level=2
+//@ run-pass
+
+// https://github.com/rust-lang/rust/issues/107975#issuecomment-1431758601
+
+use std::cell::{Ref, RefCell};
+
+fn main() {
+    let a: usize = {
+        let v = 0u8;
+        &v as *const _ as usize
+    };
+    let b: usize = {
+        let v = 0u8;
+        &v as *const _ as usize
+    };
+    let i: usize = b - a;
+
+    // A surprise tool that will help us later.
+    let arr = [
+        RefCell::new(Some(Box::new(1u8))),
+        RefCell::new(None),
+        RefCell::new(None),
+        RefCell::new(None),
+    ];
+
+    // `i` is not 0
+    assert_ne!(i, 0);
+
+    // Let's borrow the `i`-th element.
+    // If `i` is out of bounds, indexing will panic.
+    let r: Ref<Option<Box<u8>>> = arr[i].borrow();
```

Comment:
This seems to assume that the difference between addresses of the two stack variables will never be larger than 3. This is not true with cg_clif where the difference is 16.

Reply:
> This seems to assume that the difference between addresses of the two stack variables will never be larger than 3. 

Not really. It assumes that the difference is precisely 0 and non-0 at the same time. The test in question should ALWAYS panic in one place or another (even without the `assert_ne!(i, 0)`). But it doesn't, because some parts gets optimized as if `i` is 0, while others are optimized as if it is not.

What I'm trying to say is that the length choice is arbitrary and that the test will fail no matter the value of `i` (be it 0, or 1, or 3, or 16, or whatever) as long as compiler does not make contradictory assumptions during optimization.

It might actually even break if compiler still makes contradictory assumptions, but optimizes things differently.

The point of the test is to capture the *current* buggy behavior and demonstrate that it can lead to unsoundness in safe code.

Reply:
I thought the `//@ known-bug` would cause a `//@ run-pass` test like this one to fail if it doesn't panic/crash, but I guess that is not the case. Because of that I thought a non-broke backend like cg_clif is expected to not panic anywhere on this test as opposed to be supposed to panic somewhere like actually happened with cg_clif.